src/Core/Content/Product/Cms/ProductListingCmsElementResolver.php

```diff
@@ -129,21 +158,41 @@ private function addDefaultSorting(Request $request, CmsSlotEntity $slot, SalesC
                 return;
             }
 
-            $criteria = new Criteria([$sortingId]);
+            foreach ($sortCollection as $sorting) {
+                if ($sorting->getId() === $sortingId) {
+                    $customSortingKey = $sorting->getKey();
+                    break;
+                }
+            }
 
-            $request->request->set('order', $this->sortingRepository->search($criteria, $context->getContext())->first()?->get('key'));
+            if (isset($customSortingKey)) {
+                $request->request->set('order', $customSortingKey);
+            }
         }
     }
 
-    private function restrictSortings(Request $request, CmsSlotEntity $slot): void
+    /**
+     * @param EntityCollection<ProductSortingEntity> $sortCollection
+     */
+    private function restrictSortings(Request $request, CmsSlotEntity $slot, EntityCollection $sortCollection): void
     {
         $config = $slot->getTranslation('config');
 
         if (!$config || !isset($config['availableSortings']) || !isset($config['availableSortings']['value'])) {
             return;
         }
 
-        $request->request->set('availableSortings', $config['availableSortings']['value']);
+        $customSorting = new ProductSortingCollection();
+        foreach ($sortCollection as $sorting) {
+            $customSortingIds = $config['availableSortings']['value'];
+            foreach ($customSortingIds as $customSortingId => $customSortingPriority) {
+                if ($sorting->getId() === $customSortingId) {
+                    $customSorting->add($sorting);
+                }
+            }
+        }
+
+        $request->request->set('availableSortings', $customSorting->getElements());
```

Comment:
before there were only IDs under this request parameter. now you put whole sorting entities. 

that could break stuff, relying on the fact, there were only IDs before

Reply:
What could break? We pass data from private to private functions, nobody should rely on that.